ooniprobe/Utility/TestUtility.mm

```diff
@@ -86,24 +86,34 @@ + (void)downloadUrls:(void (^)(NSArray *))completion {
         path = [NSString stringWithFormat:@"%@&category_codes=%@", path, [categories componentsJoinedByString:@","]];
     }
     NSURL *url = [NSURL URLWithString:path];
-    NSURLSessionDataTask *downloadTask = [[NSURLSession sharedSession] dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
-        if (!error) {
-            NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
-            NSArray *urlsArray = [dic objectForKey:@"results"];
-            NSMutableArray *urls = [[NSMutableArray alloc] init];
-            for (NSDictionary* current in urlsArray){
-                //List for database
-                Url *url = [Url checkExistingUrl:[current objectForKey:@"url"] categoryCode:[current objectForKey:@"category_code"] countryCode:[current objectForKey:@"country_code"]];
-                //List for mk
-                [urls addObject:url.url];
-            }
-            completion(urls);
-        }
-        else {
-            // Fail
-            completion(nil);
-            NSLog(@"error : %@", error.description);
-        }
+    NSURLSessionDataTask *downloadTask = [[NSURLSession sharedSession]
+     dataTaskWithURL:url
+     completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+         if (error != nil) {
+             errorcb(error);
+             return;
+         }
+         NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
+         if (error != nil) {
+             errorcb(error);
+             return;
+         }
+         NSArray *urlsArray = [dic objectForKey:@"results"];
+         NSMutableArray *urls = [[NSMutableArray alloc] init];
+         for (NSDictionary* current in urlsArray){
+             //List for database
+             Url *url = [Url checkExistingUrl:[current objectForKey:@"url"] categoryCode:[current objectForKey:@"category_code"] countryCode:[current objectForKey:@"country_code"]];
+             //List for mk
+             [urls addObject:url.url];
+         }
+         if ([urls count] == 0){
+             errorcb([NSError errorWithDomain:@"io.ooni.api"
+                                         code:ERR_JSON_EMPTY
+                                     userInfo:@{NSLocalizedDescriptionKey:@"Error.JsonEmpty"
+                                                }]);
```

Comment:
Is it possible to factor the duplicated code used to generate this error? Also, I think it's not correct to say that the error here is an empty JSON. It's a JSON that does not contain what we expect.

Reply:
Actually it's exactly an empty JSON without urls
`[urls count] == 0`

Reply:
That's not correct. A JSON like `{"foo": "bar"}` is not empty but will cause the code to report to the caller than the error is `ERR_JSON_EMPTY`, where instead it's "error JSON structure different from expected".

Reply:
Now that I see the code, I think you should probably also validate that the fields like `"category_code"` are not empty and are indeed strings.

Reply:
And also: 

> Is it possible to factor the duplicated code used to generate this error?

Reply:
so this function requires heavy refactor:
- check url, category_code and country_code not nil, in case they are discard the input
- change ERR_JSON_EMPTY to ERR_NO_VALID_URLS or something similar, meaning that every url sent by the api is invalid

Reply:
My suggestion is to do a little bit of fake [TDD](https://it.wikipedia.org/wiki/Test_driven_development) here. We do refactor the code such that we have a function taking in input the `NSData *data`. The callback block will check whether there were errors and then call such new function. Then we write tests in which we pass data to the new function such that we provoke all the possible error conditions that we can see. We will probably see the code fail, we will learn how it is failing and decide how to repair.

Reply:
Also, `ERR_JSON_EMPTY` is probably not a description of the right error we see here.

Reply:
I want to avoid dealing with `NSData` as much as possible. But maybe I didn't understand what you want to code here. The `NSData` must be managed in the `NSURLSessionDataTask` function so it can call the relative callbacks

Reply:
I mean the following:

```ObjectiveC
+ (void)downloadUrlsCallback:(NSData *)data
                    response:(NSURLResponse *)response
                       error:(NSError *)error)
                   onSuccess:(void (^)(NSArray*))successcb
                     onError:(void (^)(NSError*))errorcb {
  // qua ci metti il codice della callback originale
}

+ (void)downloadUrls:(void (^)(NSArray*))successcb onError:(void (^)(NSError*))errorcb {
  // qua tutto come prima tranne questo
  NSURLSessionDataTask *downloadTask = [[NSURLSession sharedSession]
   dataTaskWithURL:url
   completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
     [TestUtility downloadUrlsCallback:data response:response error:error
      onSuccess:successcb onError:errorcb];
   }]];
  [downloadTask resume];
}
```

Then tests can call `downloadUrlsCallback` with specific `NSData *` that cause additional errors and we are sure that we cover all the relevant code paths.